tests/python/pants_test/backend/python/test_interpreter_cache.py

```diff
@@ -171,3 +173,32 @@ def test_setup_cached_warm(self):
   def test_setup_cached_cold(self):
     with self._setup_cache() as (cache, _):
       self.assertEqual([], list(cache._setup_cached()))
+
+  def test_interpreter_from_relpath_purges_stale_interpreter(self):
+    """
+    Simulates a stale interpreter cache and tests that _interpreter_from_relpath
+    properly detects it and removes the stale dist directory.
+
+    See https://github.com/pantsbuild/pants/issues/3416 for more info.
+    """
+    with temporary_dir() as temp_dir:
+      # Setup a interpreter distribution that we can safely mutate.
+      test_interpreter_binary = os.path.join(temp_dir, 'python2.7')
+      src = '/usr/bin/python2.7'
+      shutil.copyfile(src, test_interpreter_binary)
```

Comment:
`shutil.copy2` would be a one-stop-shop: https://docs.python.org/2.7/library/shutil.html#shutil.copy2